### PR TITLE
Upgraded loop remaining upgrades

### PR DESCRIPTION
**Pull request by nulinspiratie**
The main Upgraded loop PR (#54) was accidentally merged into master even though not all the TODO items were ticked off. Here we deal with the remaining todo items.

### Sweep
- [x] Assert that sweep_values are iterable
- [ ] Parameter.sweep automatically adds sweep to Measurement.
  Currently the awkward way is: `for k in Sweep(parameter.sweep(0,1, 0.1):`  
  Instead, `parameter.sweep` can check if it's called within a for loop, and if so, add Sweep to msmt.

### Measure
- [x] Allow measuring of a dict
- [x] Allow units to be passed when a value (not a parameter) is measured.
  Currently this is implemented but not working
- [x] Allow measuring outside of a loop (no Sweep)
  Useful if you simply want to save data once.  
  Seems to be issues with DataArrays requiring at least one set array.
- [x] Allow an empty measurement, which should be ignored

### Grouping measurements
- [ ] Create DataGroups in a Dataset that can encapsulate multiple DataArrays.
  This reflects the DataGroup of the Measurement
- [x] If a nested measurement crashes, raise an error in outer measurement

### Easy DataArray accessibility
Currently arrays in a dataset can be accessed as a key (`dataset['name_or_id']`), as an attribute `dataset.{id}` or via `dataset.get_array(s)({name_or_id})`.
We would like some more advanced and userfriendly ways to access arrays
- [ ] Easy way to access a DataArray in a DataSet (without using action_indices).
  Not yet fully figured out what the best way would be.   
  Note that it should be backward compatible
- [ ] Access measurement results within a measurement. 
  Maybe something along the lines of `msmt.get_last_measurement`.

### Metadata / snapshot
- [ ] Add metadata saving of measurement information.
  Probably the best way is to create a dict of `{action_indices: sweep/measurement}`, and have a `snapshot` method for each.
- [x] Easy access to measurement code in a dataset (`dataset.print_measurement_code()`)
- [ ] Add DataGroup information to metadata
- [ ] All metadata additions from Loop are also included in Measurement
- [x] add check that station exists before snapshotting

### Backwards compatibility
- [ ] Proper measuring of an ArrayParameter

### Step_out
- [x] Add tests ensuring functionality of `msmt.step_out`.
  In particular, step_out should work for MultiParameters etc.
- [x] Add a function such as `msmt.revert`, which allows a measurement to be repeated

### mask_attr/mask_parameter
- [x] Add `msmt.mask()`, which either directs to `mask_attr` or to  `mask_parameter`.  
  Can also accept kwargs, in which case each one is masked
- [x] Add proper tests for mask_attr/mask_parameter
- [x] Ensure a double mask is unmasked correctly
- [x] Add unmask, which can either unmask the last set value, or all previously set values
- [x] Add mask_key
- [x] Turn all mask methods into private except `msmt.mask()`
- [x] Easily mask parameters of a node (`msmt.mask(node, param1=1, param2=2)`)
- [ ] Correctly unmask if the original value is None

### Other
- [x] Fix bug where measurement still runs in a thread if it has a syntax error
- [x] Add a limit on the maximum number of arrays that can be added.
  If someone forgets to encapsulate a loop with `Sweep()`, arrays can accidentally be added nearly indefinitely
- [ ] Add support for custom setpoints when a parameter returns an array
- [ ] Ensure that measurement points are not added to the wrong data array. 
  Only matching action_indices is insufficient, there should also be a check that the parameter name matches.  
    How should we make this work if a ParameterNode/MultiParameter is measured, which then creates its own Measurement?
- [x] Add actions if measurement is completed, fail or not (`msmt.final_actions()`)
- [x] Add actions if measurement fails (`msmt.except_actions()`)
- [x] Add log message that measurement started / stopped / failed
- [x] Easily print measurement code in a dataset using `data.print_measurement()`
- [x] Add `msmt.log(message)`, which can then be picked up by logging systems
- [x] Allow `None` to be measured
- [x] Allow odd classes like `<class 'numpy.int64'>` to be measured
- [x] Add `%tb` magic to view the traceback of the last failed measurement started in a separate thread
- [x] Add notify function when measurement completes.
  Must define 
- [ ] Make `msmt.pause` a blocking function if kwarg `block=True` is provided (default)
- [ ] Do not try `data.finalize` if no arrays are created

### Optimization
- [ ] Smarter way to save metadata. 
Currently every time a new (set) array is created, the entire metadata is saved. 
In Scarlett that's a ~60 ms overhead per save
- [ ] Share set arrays between measurables

### CODE REVIEW

**Comment by maij** [comment]
Currently the way data is stored by index makes it very difficult to interpret what is stored in the data set.
```
p_get = Parameter('current
p_set = Parameter('voltage', set_cmd=None)
```
The first is a straight-forward example where there is no clear disctinction between the first and second loops in the measurement. The indexing scheme is natural and intuitive, and can't really be improved greatly, unless special indicators are otherwise added to distinguish the first and second loops (perhaps as a `Sweep` argument). One quality of life modification would be to exclude the final index if there is only 1 parameter being measured, but I'm not sure how easy that would be to do in code, especially if you have to rename the first data array when you do measure a second parameter.
```
with Measurement('m') as m:
    for k in Sweep(p_set[0:10:1]):
        m.measure(p_get)
    for k in Sweep(p_set[0:100:1]):
        m.measure(p_get)
```

```
DataSet:
   data = load_data('2020-07-14/#001_outer_14-33-41')
   <Type>   | <array_id>  | <array.name> | <array.shape>
   Setpoint | voltage_0   | voltage      | (10,)
   Measured | current_0_0 | current      | (10,)
   Setpoint | voltage_1   | voltage      | (100,)
   Measured | current_1_0 | current      | (100,)
```

---
Here's where things start to get a bit clunky. One solution to managing the way the dataset is created is to nest measurements, but seemingly very little information from the nested measurement is used to construct the dataset (e.g. in naming of data arrays). In fact, it gets worse, because even though the same sweep is executed at the 'same depth' (i.e. we're not nesting swept values), the indices change to include the depth created by the measurement.

```
with Measurement('outer') as m_outer:
    for k in Sweep(p_set[0:10:1]):
        m_outer.measure(p_get)
    with Measurement('inner') as m_inner:
        for k in Sweep(p_set[0:100:1]):
            m_inner.measure(p_get)
```
This is the current output from this example:
```
DataSet:
   data = load_data('2020-07-14/#001_outer_14-43-42')
   <Type>   | <array_id>    | <array.name> | <array.shape>
   Setpoint | voltage_0     | voltage      | (10,)
   Measured | current_0_0   | current      | (10,)
   Setpoint | voltage_1_0   | voltage      | (100,)
   Measured | current_1_0_0 | current      | (100,)
```

This is a more 'expected' output:

```
DataSet:
   data = load_data('2020-07-14/#001_outer_14-25-56')
   <Type>   | <array_id>      | <array.name> | <array.shape>
   Setpoint | voltage_0       | voltage      | (10,)
   Measured | current_0_0     | current      | (10,)
   Setpoint | voltage_1   | inner_voltage      | (100)
   Measured | current_1_0 | inner_current      | (100)
```
Note the change in indices as well as array names. I think a sprinkling of relevant info in a measurement loop can be used to construct better data array names without relying on indices (which I personally hate to look at).

What do you think of this @nulinspiratie?

**Comment by nulinspiratie** [comment]
@maij the todo items are not all implemented yet, but I think it's worthwhile to merge this.
I know it's a lot, so if you don't have the time to go through it, we can probably just merge this

File: qcodes/tests/test_measurement.py [hunk 1]
```diff
@@ -501,146 +693,690 @@ def test_double_thread_measurement(self):
...
+                raise RuntimeError('help')
+
+        self.assertEqual(p_except(), 43)
+        self.assertEqual(p_final(), 40)
+
+    def test_measurement_except_final_actions_no_fail(self):
+        p_except = Parameter(initial_value=42, set_cmd=None)
+        p_final = Parameter(initial_value=41, set_cmd=None)
+
+        with Measurement('measurement_fail') as msmt:
```
Comment (maij) [inline qcodes/tests/test_measurement.py:1348]: I think this is meant to be 'measurement_no_fail'

File: qcodes/tests/test_measurement.py [hunk 2]
```diff
@@ -501,146 +693,690 @@ def test_double_thread_measurement(self):
...
+        self.assertEqual(p_final(), 40)
+
+    def test_global_Measurement_except_final_actions_no_fail(self):
+        p_except = Parameter(initial_value=42, set_cmd=None)
+        p_final = Parameter(initial_value=41, set_cmd=None)
+
+        Measurement.except_actions.append(partial(p_except, 43))
+        Measurement.final_actions.append(partial(p_final, 40))
+
+        with Measurement('measurement_fail') as msmt:
```
Comment (maij) [inline qcodes/tests/test_measurement.py:1378]: Same with this one.

File: qcodes/tests/test_helpers.py [hunk 3]
```diff
@@ -221,10 +221,11 @@ def test_bad_calls(self):
         make_sweep(1, 3, step=1)
         # but if we change step slightly (more than the tolerance of
         # 1e-10 steps) it will fail.
-        with self.assertRaises(ValueError):
-            make_sweep(1, 3, step=1.00000001)
-        with self.assertRaises(ValueError):
-            make_sweep(1, 3, step=0.99999999)
+        # with self.assertRaises(ValueError):
```
[the remark this answers is not in the document]
Comment (nulinspiratie) [reply qcodes/tests/test_helpers.py:224]: You're right, and I actually found a serious bug, so I changed the Sweep behaviour. Good catch

File: qcodes/utils/helpers.py [hunk 4]
```diff
@@ -832,20 +833,26 @@ def _timings_to_str(self, d: dict):
 
     @contextmanager
     def record(self, key, val=None):
-        if key not in self.timings:
-            self.timings[key] = []
+        if isinstance(key, str):
+            timing_list = self.timings.setdefault(key, [])
+        elif isinstance(key, (list)):
```
[the remark this answers is not in the document]
Comment (nulinspiratie) [reply qcodes/utils/helpers.py:838]: fixed

File: qcodes/measurement.py [hunk 5]
```diff
@@ -95,7 +131,12 @@ def data_groups(self) -> Dict[Tuple[int], "Measurement"]:
     def active_action(self):
         return self.actions.get(self.action_indices, None)
 
+    @property
+    def active_action_name(self):
+        return self.action_names.get(self.action_indices, None)
+
     def __enter__(self):
+        """Operation when entering a loop"""
```
[the remark this answers is not in the document]
Comment (nulinspiratie) [reply qcodes/measurement.py:139]: okay...

File: qcodes/tests/test_measurement.py [hunk 6]
```diff
@@ -158,36 +158,161 @@ def test_new_loop_break(self):
                     if kk == 2:
                         msmt.step_out(reduce_dimension=True)
                         break
-                print('hi')
+                print("hi")
```
[the remark this answers is not in the document]
Comment (nulinspiratie) [reply qcodes/tests/test_measurement.py:161]: no, removed

File: qcodes/measurement.py [hunk 7]
```diff
@@ -159,40 +206,56 @@ def __enter__(self):
...
             return self
         except:
             # An error has occured, ensure running_measurement is cleared
             if Measurement.running_measurement is self:
                 Measurement.running_measurement = None
             raise
 
-    def __exit__(self, exc_type, exc_val, exc_tb):
+    def __exit__(self, exc_type: Exception, exc_val, exc_tb):
+        """Operation when exiting a loop
```
[the remark this answers is not in the document]
Comment (nulinspiratie) [reply qcodes/measurement.py:218]: same

File: qcodes/tests/test_measurement.py (hunk 1 above)
Comment (nulinspiratie) [reply qcodes/tests/test_measurement.py:1347]: yep

File: qcodes/tests/test_measurement.py [hunk 8]
```diff
@@ -158,36 +158,161 @@ def test_new_loop_break(self):
...
-    #         self.assertEqual(msmt.loop_shape, ())
-    #         msmt.measure(self.p_measure)
+    def test_new_loop_0D(self):
+        # TODO Does not work yet
+        with Measurement("new_loop_0D") as msmt:
+            self.assertEqual(msmt.loop_shape, ())
+            msmt.measure(self.p_measure)
+
+    def test_new_loop_1D_0D(self):
+        # TODO Does not work yet
```
[the remark this answers is not in the document]
Comment (nulinspiratie) [reply qcodes/tests/test_measurement.py:190]: yep

**Review by maij — APPROVED** [review]

**Pull request merged**